### ISSUE TITLE
cafemocha:test aborted

### ISSUE THREAD

**Issue by alsoicode** [AndrewKeig/advanced-express-application-development#3]
Hello,

I'm working through Chapter 1, and receiving the following error when running:

```
$ grunt test

Running "env:test" (env) task

Running "cafemocha:test" (cafemocha) task
Warning: Cannot log to file without filename or stream. Use --force to continue.

Aborted due to warnings.
```

I have copy/pasted the gruntfile.js from the source code here. What am I doing wrong or missing?

Thanks!

**Comment by AndrewKeig** [comment]
Hi

You need a config/test.json file: 

{
  "express": {
    "port": 3000
  },
  "logger" : {
    "filename": "logs/run.log",
    "level": "silly"
  }
}

**Comment by alsoicode** [comment]
Hi Andrew,

Thanks for the reply. I actually do have config/test.json as specified.

My directory structure is:

```
vision/
    config/
        coverage.json
        default.json
        development.json
        test.json
```

**Comment by AndrewKeig** [comment]
Cool; the only way I could replicate the error was with this in the config file:

As you can see the filename is empty; this throws the same error.

Can you setup a repo for this so I can take a look.

{
"express": {
"port": 3000
},
"logger" : {
"filename": "",
"level": "silly"
}
}

**Comment by alsoicode** [comment]
Ok, my progress so far is at: https://github.com/alsoicode/vision

**Comment by AndrewKeig** [comment]
Hopefully this should also resolve.  Let me know if you still have a problem.

**Comment by alsoicode** [comment]
Hmm. Running:

```
mocha test
```

will work just fine, but I still get the same error from Grunt. Is it not able to fund the log file or something?

**Comment by AndrewKeig** [comment]
Lets see what the console.log says; let me know if anything comes through.  I will test this on linux later.

Try this in ../logger/index.js to print out the file path

```
var winston = require('winston'),
    config = require('../configuration');

function Logger() {
    console.log('filename', config.get("logger:filename"));

    return winston.add(winston.transports.File, {
        filename: config.get("logger:filename"),
        maxsize: 1048576,
        maxFiles: 3,
        level: config.get("logger:level")
    });
}

module.exports = new Logger();
```

For now you can simply switch on the console logger; replace ../logger/index.js like so:

```
function Logger() {
    return winston;
}

module.exports = new Logger();
```

**Comment by alsoicode** [comment]
I get:

```
filename undefined
```

from the `console.log()` statement

**Comment by AndrewKeig** [comment]
It seems config.get is returning undefined; which means we cannot find the item "logger:filename" inside config.  I would suggest replacing the ./lib/configuration/index.js with this; let me know what values you get for environment/key.

```
var nconf = require('nconf');

function Config() {
    nconf.argv().env('_');
    var environment = nconf.get('NODE:ENV') || 'development';
    console.log("environment: ", environment);
    nconf.file(environment, 'config/' + environment + '.json');
    nconf.file('default', 'config/default.json');
}

Config.prototype.get = function(key) {
    console.log("key: ", key);
    return nconf.get(key);
}

module.exports = new Config();
```

**Comment by alsoicode** [comment]
Finally figured out what the problem was. `environment` is returned as uppercase: "TEST", so when the logger tries to open `config/TEST.json`, it's not there, but `config/test.json` is.

So, I just changed the environment to be returned as lower case:

```
var nconf = require('nconf');

function Config() {
    nconf.argv().env('_');
    var environment = nconf.get('NODE:ENV').toLowerCase() || 'development';
    nconf.file(environment, 'config/' + environment + '.json');
    nconf.file('default', 'config/default.json');
}
```

and now:

```
$ mocha test
```

will pass :) Yeah!

**Comment by AndrewKeig** [comment]
There is no need to make this change; why have you called the file TEST.json; its very clear in the book on page 13; 

"Now add the following configuration to the development, test, and coverage config files: ./config/development.json, ./config/test.json, and ./config/ coverage.json."

**Comment by alsoicode** [comment]
Hi,

Actually, it's the other way around. I named the file test.json - lowercase, but the environment variable was coming back uppercase for me, hence the need to take it to lower case in order to find the file. I'm triple-checking my code against the book, and trying to be very careful to follow the examples to the letter.

If you look at my source code, you'll see that the file is named lowercase as well.

Brandon

**Comment by AndrewKeig** [comment]
Hi Brandon

Apologies; you are right; Linux is not so forgiving on filenames; which is why it passed on my mac.

If you would like to contribute a fix; please do.

**Comment by alsoicode** [comment]
No worries at all. By the way, I revised my review on Amazon to be more constructive, so my apologies for getting frustrated as well. I gave you 4 stars; it's a well written book. I'm primarily a Python developer, so JavaScript isn't my strongest suit, and I tend to not have as much patience with it, or even know where to look for problems sometimes, but I'm very anxious to learn more, which is why I bought your book.
